### PR TITLE
Add masked_select abstract impl

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110103

Fixes https://github.com/pytorch/pytorch/issues/109871

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -593,6 +593,28 @@ def nonzero(fake_mode, func, arg):
     return arg.new_empty((arg.nonzero_memo, arg.dim()), dtype=torch.int64)
 
 
+@register_op_impl(lambda func: func is torch.ops.aten.masked_select.default)
+def masked_select(fake_mode, func, self, mask):
+    if (
+        fake_mode.shape_env is None
+        or not fake_mode.shape_env.allow_dynamic_output_shape_ops
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:600]: I forget why we have this config (as opposed to dyn output shapes always being allowed when we have a shape env), but i'll read the code later - see we do the same for nonzero.

File: torch/_subclasses/fake_tensor.py [hunk 2]
```diff
@@ -593,6 +593,28 @@ def nonzero(fake_mode, func, arg):
...
+        # Without symints/symfloats, cannot handle this
+        raise DynamicOutputShapeException(func)
+
+    nnz = fake_mode.shape_env.create_unbacked_symint()
+
+    # see nonzero for commentary
+    maxval = sys.maxsize - 1
+    if not free_symbols(arg.numel()):
+        if arg.numel() >= 2:
+            maxval = int(arg.numel())
```
Comment (bdhirsh) [inline torch/_subclasses/fake_tensor.py:611]: i wonder if this pattern will become common enough that we should util-ify it

**Review by bdhirsh — APPROVED** [review]

File: torch/_subclasses/fake_tensor.py (hunk 1 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:600]: The reason is inductor currently chokes on this stuff, so we need to induce a graph break under default config. But https://github.com/pytorch/pytorch/pull/109893 should make it so we can start attempting to change the defaults here.

File: torch/_subclasses/fake_tensor.py (hunk 2 above)
Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:611]: There's definitely some design that needs to be thought through for upper bounds on symbolic sizes. We actually probably do want to track upper bounds that have symbolic symbols in them, for other reasons...

**Pull request closed without merging**